### ISSUE TITLE
Tile Size per layer not per Map

### ISSUE THREAD

**Issue by RolandasRazma** [mapeditor/tiled#149]
I think "Tile Size" should be moved to layer property instead of map. It makes way more sense and enables to use tiled way more flexible.

**Comment by Ulli-RPG** [comment]
I fully support this request, tile size should be per layer, not per map. Would be wonderful to have it implemented!
(I can't find the reference option, here's another dead end thread with the same request: https://github.com/mapbox/mapbox-android-sdk/issues/151)

**Comment by bjorn** [comment]
@Ulli-RPG I already said in my first comment that I agree doing tile size per layer would be a good idea. It's just a matter of getting around to it (and solving the related problems since it won't be a trivial patch).

I'm not sure what you meant to achieve by linking to some seemingly related but totally irrelevant issue on the mapbox project, though...

**Comment by Ulli-RPG** [comment]
Sorry, I'm new to this all stuff, I just wanted to say that I also support this idea.
For the link I thought it was related, overlooked it was another project / Android.

**Comment by bjorn** [comment]
@Ulli-RPG Alright. :)

**Comment by abueide** [comment]
@bjorn how would we get a bounty going for this? I'd chip in

**Comment by bjorn** [comment]
@abueide You can set a bounty on this issue at https://www.bountysource.com/issues/128123-tile-size-per-layer-not-per-map. Usually bounties are quite successful in attracting contributors, though I'm afraid this issue will be a tough one to resolve. But we'll see. I hope to eventually also get around to it of course. Thanks in advance for your support!

**Comment by Lucrecious** [comment]
Hey @bjorn 

So I think I'm going to try and add in the tile per layer support (if no one else is going to try it), but there are a few caveats.

Selection between layers with different tile size will dissipate (I don't mind losing out on this functionality - of course this will be open source, so if anyone wants to tackle it, they can).

All layers in an automapping file need to have the same tile size (in my opinion, if the input and output layer are different tile sizes, the automapping is undefined anyways). So basically, when an automapping is ran on a file with layers of different tile sizes, you'll get an error (just like if you were to omit having regions_input or regions_output).

I'm wondering if I should know about any more glaring issues you might have thought about between 2012 and now, because that would help a lot. It's going to take a while for me to figure out the chain of discrepancies for the map tileHeight and tileWidth.

Thanks, and great app.

**Comment by bjorn** [comment]
@Lucrecious Please open a pull request with that branch even if it's not entirely done yet. Then we can discuss the changes there. Thanks for taking on this issue!

**Comment by Lucrecious** [comment]
Here's a the pull request: #1422 
I spent some time on documenting what I've done so it's easier to continue!

**Comment by Gr33nbl00d** [comment]
Really would like this feature

**Comment by Lucrecious** [comment]
Here's my PR for it: #1422 
Once I have more time, I'll be updating it since my edits' pretty messed up right now. I'm currently working on a project of my own, so it'll be a while until I can get back to this. 

As of now, you can use my fork of Tiled, which includes these changes (although very buggy). You can also do the changes yourself!

**Comment by Gr33nbl00d** [comment]
If i had the time i would do it but last time i wrote c/qt is about 8 years ago :) Java javascript or dotnet would be no problem ;) but as im also busy with my project it would take too much time get into it again... i would enjoy it but for the moment i did a workarround rendering 2 different tile maps instead of one. But if i get back to individual level design and its not done by than, i will give it a try :)

**Comment by Lucrecious** [comment]
@Gr33nbl00d  #1422 is my fork which already has tile size per layer. It's just a little buggy, but I use it just fine. Try it out when you get the chance, all you have to do is clone and build.

**Comment by Gr33nbl00d** [comment]
Ok i give it a try, would make things much easier for me :)

Am 29.06.2017 9:30 nachm. schrieb "Luca Duran" <notifications@github.com>:

**Comment by bjorn** [comment]
Feature [requested on the forum](http://discourse.mapeditor.org/t/layers-with-different-tile-sizes/3120).

**Comment by fmscorpion** [comment]
I think I have an idea that might be good for working with different layers, that is improving the object layer, I mean, if there was a snap feature on the object layer when placing tiles it would "solve" the issue of working with different layer sizes, if, the layer sizes are multiples. An example is if I had an 64x64 layer tile size and I use the object layer with 32x32 layer size it would snap four tiles of 32x32 inside the region of 64x64. Fortunatelly it would "solve" this problem without needing to make major alterations on the code

**Comment by bjorn** [comment]
@fmscorpion Well, snapping objects to the tile grid, as well as snapping them to a subdivided grid, is already available. See _View > Snapping_ menu and _Edit > Preferences > Interface > Fine grid divisions_.

**Comment by fmscorpion** [comment]
I didn knew this feature but it works very well, by the way, I realized something that could be updated, maybe multiple placings could be implemented in the objects layer, I mean, it works, but it sometimes looks like a lot of work to place tile by tile even with the snap on.

**Comment by HollowedEmpire** [comment]
I would also like this feature. I have three different sizes for tiles that have different usage purposes. I have had to use awkward workarounds varying from multiple maps to leaving large gaps in tile layers that use a large tiles and making hacky exporters skip over these gaps depending on the layer in question.

**Comment by JamesO2x** [comment]
I just had a thought on this feature. I think a simple solution would be to only allow layers use tile sizes that are multiples of the base tile size defined in the "Map" properties.

For example, if your map properties use 32x32 pixel tiles as default, you could define a property in another layer that is called "Tile Layer Multiplier". When "Tile Layer Multiplier" is set to "2" it would mean that the tile size for that layer is "64x64" pixels (a 2 multiplier of the base tile size). If you set the multiplier to 10, you could have a tile sizes of 320x320.

I think this is a reasonable implementation that shouldn't be too complex in the math/implementation side of things.

I understand that the ideal solution would be to allow exact pixel size definitions, but realistically... I can't imagine **any** real world use-case where someone would want to have (for example) a layer with 32x32 tiles, and another layer with something like _31x31_ tiles. It would be a nightmare to design a map like that, with minor offsets that don't line up nicely.

I would think that such a truly arbitrary tile-size definition would actually cause **more** headache for map designers than it is worth, because it would promote bad design practice. Unless someone were to come forward with a **very compelling argument** on why different layers should allow tiles that don't evenly divide into each other...

I think using a multiplier approach would not only be the simplest solution for both programmers and map designers, I think its also the most sensible approach that would cover 99% of the cases where someone would need differently sized tile layers.

That said, using this "multiplier" approach would also still "technically" allow for fully arbitrary tile sizes. For example, if you set your Map's default tile size to "1x1" then you could make a new layer that is any arbitrary size of pixel multiplier. You could make a cursed map with 32x32 and 31x31 tiles, if you really wanted to.

**Comment by bjorn** [comment]
> That said, using this "multiplier" approach would also still "technically" allow for fully arbitrary tile sizes. For example, if you set your Map's default tile size to "1x1" then you could make a new layer that is any arbitrary size of pixel multiplier. You could make a cursed map with 32x32 and 31x31 tiles, if you really wanted to.

I think this kind of reveals a problem with the proposal, because it means the implementation is actually not any easier than an implementation that can handle arbitrary tile sizes. It might be that an implementation that handles only power-of-two multipliers (so x2, x4, x16, etc.) could be easier, but even that leaves the problem that a tile selection of some layer can include partial tiles from another layer.

In any case, thank you for considering ways to make this easier! Maybe there is a way, but I'm afraid it's going to be a major project either way where my initial comments from 2012 highlight the pain points.

**Comment by JamesO2x** [comment]
I think my proposal is still easier than one that handles truly arbitrary tile sizes though. Consider the following:

If you attempt to use truly arbitrary tile sizes, then there needs to be some function within Tiled that takes the total number of layers in a map, finds some common denominator between all of them that is divisible by all of them, and then adjust layers grids and such accordingly. Then, every time a new layer is created, this same computation has to be done all over again. Multi-layer selection would be a real nightmare to program I would think.

With my proposal, you would never be able to create new layers with tiles _smaller_ than the map's default size. This basic assumption simplifies a lot of the math. since all future layers must be perfect multiples of this base value, Selection matrices merely have to be multiplied/divided by whole integers (2, or 3 or 4, etc) when moving between layers. And yes you're right, you could probably simplify this further to limiting the multiplier to powers of 2.

Basically, you simplify the solution by simplifying the problem. I really don't think anyone actually needs a single map file with tiles that are 13 pixels, 458 pixels, 2 pixels, 3 pixels, 16 pixels, and 17 pixels all at the same time. So the "problem" of true arbitrary tile sizes is a bit of a silly one... its like trying to invent a machine that sorts the hair on your head by length - sure you _can_ create such a machine if you really want to... but is that really a problem that any actual person needs solving?

To handle the "partial tiles" problem, I think a design decision just needs to be made: Either round the selection up to the nearest full tile, or round it down.. I don't think it needs to be any more complicated than that.

**Comment by MarkOates** [comment]
👋 I would like this feature as well (and would prefer full arbitrary tile sizes, not power of two).

> but even that leaves the problem that a tile selection of some layer can include partial tiles from another layer

I'm OK with that! My initial solution for this problem was to simply *not* provide support for multilayer selection, in cases where the map has different tile sizes on layers.  A step forward from that would be to activate layers in the layer panel that you would want to select through, possibly with an icon of sorts.

I also like the idea of orientation and transformation being layer dependent (interesting platforming on rotated maps!). In a perfect world, you could have any layer (of any type) with unique tile sizes, orientation, and transformation.

**Comment by JamesO2x** [comment]
> 👋 I would like this feature as well (and would prefer full arbitrary tile sizes, not power of two).

Can you describe your use case for fully arbitrary tile sizes per layer? Describing your use case should help the programmers understand how people are using this tool (and how they might want to use it), which can help with implementation.

Also I'm genuinely curious. Because I've seen arbitrary tile sizes requested before, but I've never seen anyone actually describe a use case that would actually require truly arbitrary tile sizes per layer (that couldn't be solved with proper fore-planning).

For example, I once had a freelance client who insisted that the tiles in his project must be 23 pixels wide, even after I explained that it would leave gaps in the memory when loaded into the engine, because the tile table would likely round up reserved space to 24 or 32 pixels anyway. Sure, modern computers nowadays have enough RAM to spare, it won't tank performance, but it's also an unnecessary waste when adding one extra column of pixels would not only take advantage of the wasted space, but also fit neater onto normal screen resolutions, and be more in line with industry standard.

---

As for my own use case, I want to be able to do something akin to Donkey Kong Country or the Genesis Sonic games, where the levels are made up whole-sceeen sized "chunks" (called meta-tiles, usually 256x256 px) while the normal sprites and objects are smaller tiles of 16x16 px. The larger tiles are snapped to their own large grid, but are e a multiple of the smaller tiles, so everything aligns nicely.

Same for DKC, but I believe their tile size was 24 px. _Technically_ in DKC the levels' larger meta-tiles are made up of small 24x24 tiles, but when designing the game the devs weren't actually placing these 24x24 tiles manually ever. The 3D level chunks were designed and rendered as 256 px images, and then later sliced up into memory as 24 px tiles as a hardware limitation convenience.

But modern gaming hardware wouldn't likely struggle with actual full 256 px tiles. So a modern workflow would likely have level geometry on its own 256 px tile grid layer, with a smaller 24 px tiles layer to add individual blocks for adjustment, bananas, coins, etc.

My use case would never require arbitrary tile sizes. Any layer would always be a multiple of the smallest base tile unit.

**Comment by Snowlith** [comment]
Just want to mention that with Godot's change from TileMap nodes to TileMapLayer nodes in the upcoming 4.3 release, the functionality of determining tile size per layer instead of map would be useful for Godot users.

**Comment by bjorn** [comment]
@Snowlith Oh, that's interesting. Do you know how they deal with multi-layer editing? I read you can select multiple TileMapLayer nodes at once, but I wonder what happens if you try to to select and edit with multiple selected layers that have different tile sizes.

**Comment by Snowlith** [comment]
@bjorn The Godot editor does not allow editing of multiple layers at once, even layers of the same tile size. The Godot tile editor is definitely not as complete as Tiled's editing capabilities, and I feel like Tiled is currently serving as a valuable reference for the many devs working on these features 😄 
![image](https://github.com/mapeditor/tiled/assets/158532337/8cae8a0c-ed8f-403d-9ac7-eb3fbd32f7fc)
Good luck solving this difficult problem!
